### ISSUE TITLE
nn.init.sparse poor performance

### ISSUE THREAD

**Issue by stefanonardo** [pytorch/pytorch#6021]
- OS: Ubuntu 16.04
- PyTorch version: 0.3.1
- How you installed PyTorch (conda, pip, source): conda
- Python version: 3.6.4
- CUDA/cuDNN version: 9.1 
- GPU models and configuration: GTX 1060

It seems that `nn.init.sparse` takes too much time for its purpose. On my CPU, a 500x500 tensor initialization takes 0.25 seconds. 
I wrote a piece of code that can be 10x faster than `nn.init.sparse`. I would ask you to try it yourself on your machines and check whether the semantic is the same.

```python
import torch
import torch.nn as nn
import time

sparsity = 0.9
tensor_size = 500

t1 = torch.Tensor(tensor_size, tensor_size)
s = time.time()
nn.init.sparse(t1, sparsity)
print(time.time() - s)

t2 = torch.Tensor(tensor_size * tensor_size)
s = time.time()
t2.normal_()
if sparsity > 0:
    zero_weights = torch.randperm(int(tensor_size * tensor_size))
    zero_weights = zero_weights[:round(tensor_size * tensor_size * sparsity)]
    t2[zero_weights] = 0
t2 = t2.view(tensor_size, tensor_size)
print(time.time() - s)
```

**Comment by stefanonardo** [comment]
I would just like to point out that Bernoulli implementation has a different semantic. In the worst case, the actual tensor's sparsity could be much different from the function's parameter.

**Comment by fmassa** [comment]
Actually, I believe both implementations are slightly different than the one we current have.

Indeed, the current implementation guarantees that we have a specific number of non-zero elements **per row**, while the first implementation proposed here only guarantees a fixed number of non-zero elements **per tensor**. 

I have never used this weight initialization, so I don't know if this makes a difference in practice.

**Comment by mttk** [comment]
@fmassa correct, that crossed my mind now as well. An option would be to allow the user to select row-level stability (but a slower speed) or tensor-level stability (faster). 

The Bernoulli option breaks all of this, so I'm not as much in favor of it anymore after getting temporarily blinded by its simplicity. I crunched some numbers and the sparsity does not deviate a lot when using it, but it does deviate.

Guess it depends which way you guys prefer it, and also what the use cases are.

**Comment by fmassa** [comment]
another option could be something like (not sure if slower, and I didn't test it):
```
_, idxs = torch.rand(tensor.size(), dtype=tensor.dtype).topk(num_zeros, 1)
t[idxs] = 0
```

**Comment by fmassa** [comment]
@stefanonardo @mttk would any of you have time to send a PR improving the performance of `nn.init.sparse_`?

**Comment by mttk** [comment]
@fmassa yeah, sure. I'll write a shootout between the last two options. @stefanonardo let me know if you want to do it since this looks like it's going to mostly be your code.

**Comment by stefanonardo** [comment]
No problem, you can do it.

**Comment by fmassa** [comment]
Fixed in #6899
